### PR TITLE
Add containers test

### PR DESCRIPTION
**Pull request by bitoku**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR adds unit tests to crictl (a part of container.go).
This also adds stretchr/testify and go-test/deep library.

#### Which issue(s) this PR fixes:

#1152

#### Special notes for your reviewer:

I wrote tests with default unittest framework.
Which do you prefer to use test framework, ginkgo or default?
Exisiting unittests in cmd/crictl is written with the default one.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: cmd/crictl/container_test.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+			opts: listOptions{
+				last: 7,
+			},
+			expect: containers,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			actual := getContainersList(containers, tc.opts)
+			deep.Equal(tc.expect, actual)
```
Comment (SergeyKanzhelev) [inline cmd/crictl/container_test.go:68]: Why not `reflect.DeepEqual`? Eliminates external dependency

File: cmd/crictl/container_test.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"k8s.io/kubernetes/pkg/kubelet/types"
+	"testing"
+	"time"
+
+	"github.com/go-test/deep"
+	"github.com/stretchr/testify/assert"
```
Comment (SergeyKanzhelev) [inline cmd/crictl/container_test.go:9]: I see it is used in k/k in places. But I am not sure how badly this repo needs another dependency

File: cmd/crictl/container_test.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+func TestGetPodNameFromLabels(t *testing.T) {
+	testCases := []struct {
+		desc   string
+		input  map[string]string
+		expect string
+	}{
+		{
+			"pod name label should be returned",
+			map[string]string{types.KubernetesPodNameLabel: "test"},
```
Comment (SergeyKanzhelev) [inline cmd/crictl/container_test.go:81]: maybe also test cases with empty map and more than one label

File: cmd/crictl/container_test.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	"testing"
+	"time"
+
+	"github.com/go-test/deep"
+	"github.com/stretchr/testify/assert"
+	pb "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+// fakeContainers creates fake containers in the least recent order of the createdAt.
+func fakeContainers(names ...string) []*pb.Container {
```
Comment (SergeyKanzhelev) [inline cmd/crictl/container_test.go:14]: rename to `fakeContainersWithCreationTimeAsc` or something

**Review by saschagrunert — CHANGES_REQUESTED** [review]
Thank you for working on this, @bitoku! :pray:

File: cmd/crictl/container_test.go (hunk 2 above)
Comment (bitoku) [reply cmd/crictl/container_test.go]: ok I'll make it use ginkgo

**Comment by bitoku** [comment]
There seems no CI for crictl so I added them as well.
Thank you for reviewing!

File: cmd/crictl/container_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+				Name: name,
+			},
+			CreatedAt: creationTime.UnixNano(),
+		})
+		creationTime = creationTime.Add(time.Second)
+	}
+	return containers
+}
+
+var _ = Describe("container.go", func() {
```
Comment (SergeyKanzhelev) [inline cmd/crictl/container_test.go:45]: do you need a file name describe? Sounds excessive, should be clear from this file name.

**Comment by bitoku** [comment]
sorry for the late reaction.
I reflected your comment. could you please review it again?

**Comment by saschagrunert** [comment]
@SergeyKanzhelev PTAL again

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Review by saschagrunert — APPROVED** [review]

**Comment by bitoku** [comment]
I'm sorry for bothering you many times.
I tested it locally and it worked this time. not sure why it failed because it looks ok. need to check.

https://github.com/kubernetes-sigs/cri-tools/actions/runs/6874923452/job/18697587316?pr=1194

**Comment by saschagrunert** [comment]
I don't get why we're getting an exit code 1 for the succeeding test suite, too. :thinking:

**Comment by SergeyKanzhelev** [comment]
Couple options to debug - increase the ginkgo log level and try removing the `-p` as a test. There is definitely something happens. Maybe a panic or something that is not being logged

**Comment by bitoku** [comment]
After some research I found the reason why it failed.

First of all, the reason it didn't happen in my local env was that I didn't rebase it onto upstream/main. After rebasing it, I was able to reproduce it.

And the CI failed because ginkgo somehow runs `templates_test.go`, which doesn't use ginkgo, and its test failed.
I don't know why ginkgo runs no-ginkgo spec file. I raised a issue about it (https://github.com/onsi/ginkgo/issues/1303).

`templates_test.go` failed because of this change https://github.com/kubernetes-sigs/cri-tools/commit/092eddc1769d54ab9da2c589d6449d2256a1eece#diff-911ebdfd44eef4e1d8f910ef1839b3e99c09a2fc4da1eb8b8f464e5bbfcafe9c .

`strings.Title` and `cases.Title` behave differently. To match `cases.Title` with `strings.Title`, it needs `cases.NoLower` as its option.
https://go.dev/play/p/jDL6uoPVB28

I tested in my local and my forked repo, and it worked so I believe it works this time...

**Review by saschagrunert — APPROVED** [review]
